File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        def long(X, Y, m, n, seen):
                k = (m,n)
                if m == 0 or n == 0:
                    return 0
                if k not in seen:

                    if X[m-1] == Y[n-1]:
                        seen[k] = long(X, Y, m-1, n-1, seen) + 1
                    else:
                        a = long(X, Y, m-1, n, seen)
                        b = long(X, Y, m, n-1, seen)
                        seen[k] = max(a, b)
                return seen[k]

        x = len(text1) 
        y = len(text2)
        seen = {}
        return long(text1, text2, x, y, seen)

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int l1=text1.length(), l2=text2.length();
        vector<vector<int>> c(l1+1, vector<int>(l2+1, 0));
        for(int i=0;i<l1;i++)
            for(int j=0;j<l2;j++)
             if(text1[i]==text2[j]) c[i+1][j+1]=1+c[i][j];
                else c[i+1][j+1]=max(c[i+1][j], c[i][j+1]);

            return c[l1][l2];
    }
};